content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
+
+Flatpak can be used during development time with either [GNOME Builder](https://flathub.org/apps/details/org.gnome.Builder), VSCode with the [flatpak-vscode](https://marketplace.visualstudio.com/items?itemName=bilelmoussaoui.flatpak-vscode) extension or [fenv](https://gitlab.gnome.org/ZanderBrown/fenv/) if you prefer the CLI for building/running your application.
+
+- <https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-stable>
+- <https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-nightly>
+
+## Portals
+
+Modern Linux applications should make use of the [portals](https://flatpak.github.io/xdg-desktop-portal/) to request runtime permissions to access resources such as the camera, starting a screencast session or picking a file. 
+
+[ASHPD](https://crates.io/crates/ashpd) is nowadays the way to go for using portals from Rust as it provides a convenient & idomatic API on top of the DBus one.
+
+Code example from `ashpd-0.4.0-alpha.1` for using the color picker in a desktop environment agnostic way.
+```rust
+use ashpd::desktop::screenshot::ColorResponse;
+
+async fn run() -> ashpd::Result<()> {
+    let color = ColorResponse::builder().build().await?;
+    println!("({}, {}, {})", color.red(), color.green(), color.blue());
+    Ok(())
+}
+```
+
+
+## Keyring & oo7
+
+Applications that needs to store sensitive information such as passwords usually go through the [Secret service](https://specifications.freedesktop.org/secret-service/latest/) which is a protocol supported by both gnome-keyring and kwallet nowadays. There were multiple attempts to provide a Rust wrapper of the DBus API but sadly those were lacking either async support, they didn't provide integration with the secret portal or had no way to allow applications to migrate their secrets from the host keyring to the application sandboxed one.
+
+Those were the primary reasons we started working on [oo7](https://crates.io/crates/oo7), it is still on alpha stage, but should cover most of the use cases already.
+
+## GStreamer
+
+As part of the experiments I did when porting [Authenticator](https://gitlab.gnome.org/World/Authenticator) to GTK 4 was figuring out how can I replicate the internal GStreamer sink that GTK has to convert a video frame into a `gdk::MemoryTexture` in order to render it in some widget for QR codes scanning purposes.
+
+Jordan Petridis took over my WIP work and turned it into a proper GStreamer plugin written in Rust, see <https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/tree/main/video/gtk4> for an example on how to use it in your application.
+
+## Missing integrations
+
+### Better gettext support
+
+Currently, [gettext](https://www.gnu.org/software/gettext/) doesn't support Rust officially and most importantly, it doesn't like the `!` character, so we can't use string formatting for translatable texts. 
```

Comment:
Just like this nobody (who doesn't know yet) will understand what the relation between ! and not being able to translate strings is :)

Reply:
I have added a sentence explaining that `!` is used in declarative macros such as `format!` / `println!`